### PR TITLE
[0.4.0] Implement compound assignments (+= and friends)

### PR DESCRIPTION
**Pull request by ruby0x1**
So I got this working well with fairly low code impact!
Posting a PR as is so @munificent and others can have a look and people can mess with it in the mean time.

~~**todo: add tests.**~~
~~I've got a big change locally that's making that harder for me right now, 
if someone wants to dive into adding tests feel free.~~

**Don't merge this yet**. 
This is a 0.4.0 change.

**Example:**
devenv_M0RaX9Zfpd

**Benchmarks with master as baseline**
image

### CODE REVIEW

File: src/vm/wren_compiler.c [hunk 1]
```diff
@@ -2695,6 +2727,92 @@ void expression(Compiler* compiler)
...
+}
+
+static void setterAssignment(Compiler* compiler, Code instruction, Signature signature)
+{
+  TokenType infixToken = compoundAssignmentStart(compiler);
+  if(infixToken == TOKEN_EQ) return;
+
+  //We're doing self.setter = self.getter + ...
+  //so to call the getter, we duplicate the self on the stack
+  emitOp(compiler, CODE_LOAD_LOCAL_0);
```
Comment (underscorediscovery) [inline src/vm/wren_compiler.c]: This fails when 0 is not self (i.e inside a Fn closure). It should basically be stack -1, will revise.

**Comment by mhermier** [comment]
For the record, the implementation doesn't provide all the binary
operators. So as this it is a no go for me.

**Comment by underscorediscovery** [comment]
one step at a time @mhermier.

**Comment by mhermier** [comment]
Please cleanup the change set, it starts to become unreadable.
Introduce the new opcode in a single patch, and rename it. It is not a PUSH, it is part of the DUP familly, something like DUP_CALLFRAME.
Also try to make commits a little bit more atomic, for ease of readability and future maintenance.
Thanks.

**Comment by bjorn** [comment]
> Please cleanup the change set, it starts to become unreadable.

@mhermier I believe those commits are just his WIP. We can look at the entire patch by clicking on "Files changed", which I think is still relatively small. Personally I'd squash it all into a single commit when merging.

**Comment by mhermier** [comment]
From my past experience I prefer to separate pure vm changes from languages
ones at minimum. It makes it easier to bissect errors. Likewise if entries
are more atomic.

**Comment by mattiasljungstrom** [comment]
If you're doing this, how about adding a++, and a-- as well? Should be doable with this new op code? I'm missing a++ much more than a += 4 when writing code in Wren.

**Comment by ruby0x1** [comment]
Going to reopen a new PR soon with a better implementation (one that works, this one doesn't work safely and will mess up the stack).

**Pull request closed without merging**